applications/OptimizationApplication/custom_utilities/response/linear_strain_energy_response_utils.cpp

```diff
@@ -0,0 +1,311 @@
+
+    // calculate sensitivities for each and every model part w.r.t. their sensitivity variables list
+    for (const auto& it : rSensitivityModelPartVariableInfo) {
+        auto& r_sensitivity_model_part = *(it.first);
+        for (auto& r_variable : it.second) {
+            std::visit([&](auto&& r_variable) {
+                if (*r_variable == YOUNG_MODULUS_SENSITIVITY) {
+                    CalculateStrainEnergyYoungModulusSensitivity(r_sensitivity_model_part, YOUNG_MODULUS_SENSITIVITY);
+                } else if (*r_variable == POISSON_RATIO_SENSITIVITY) {
+                    CalculateStrainEnergyNonLinearSensitivity(r_sensitivity_model_part, PerturbationSize, POISSON_RATIO, POISSON_RATIO_SENSITIVITY);
+                } else if (*r_variable == SHAPE_SENSITIVITY) {
+                    CalculateStrainEnergyShapeSensitivity(r_sensitivity_model_part, PerturbationSize, SHAPE_SENSITIVITY);
+                } else {
+                    KRATOS_ERROR
+                        << "Unsupported sensitivity w.r.t. " << r_variable->Name()
+                        << " requested for " << r_sensitivity_model_part.FullName()
+                        << ". Followings are supported sensitivity variables:"
+                        << "\n\t" << YOUNG_MODULUS_SENSITIVITY.Name()
+                        << "\n\t" << POISSON_RATIO_SENSITIVITY.Name()
+                        << "\n\t" << SHAPE_SENSITIVITY.Name();
+                }
+            }, r_variable);
+        }
+    }
+
+    KRATOS_CATCH("");
+}
+
+
+void LinearStrainEnergyResponseUtils::CalculateStrainEnergyShapeSensitivity(
+    ModelPart& rModelPart,
+    const double Delta,
+    const Variable<array_1d<double, 3>>& rOutputSensitivityVariable)
+{
+    KRATOS_TRY
+
+    using tls_type = std::tuple<Vector, Vector, Vector, Element::Pointer>;
+
+    const auto& r_process_info = rModelPart.GetProcessInfo();
+
+    const int max_id = block_for_each<MaxReduction<int>>(rModelPart.GetRootModelPart().Nodes(), [](const auto& rNode) {
+        return rNode.Id();
+    });
+
+    std::vector<std::string> model_part_names;
+
+    block_for_each(rModelPart.Elements(), tls_type(), [&](auto& rElement, tls_type& rTLS) {
+        if ((!rElement.IsDefined(ACTIVE) || (rElement.Is(ACTIVE))) && rElement.Is(SELECTED)) {
+            Vector& r_u = std::get<0>(rTLS);
+            Vector& r_ref_rhs = std::get<1>(rTLS);
+            Vector& r_perturbed_rhs = std::get<2>(rTLS);
+            Element::Pointer& p_element = std::get<3>(rTLS);
+
+            auto& r_geometry = rElement.GetGeometry();
+            const auto domain_size = r_geometry.WorkingSpaceDimension();
+
+            rElement.GetValuesVector(r_u);
+
+            // calculate the reference value
+            rElement.CalculateRightHandSide(r_ref_rhs, r_process_info);
+
+            // initialize dummy element for parallelized perturbation based sensitivity calculation
+            // in each thread seperately
+            if (!p_element) {
```

Comment:
In this implementation, we are basically copying the model part mesh and deleting it later, which is a big memory overhead. We could create local points and elements to perform the calculation without the need for creating model parts and respectively nodes, etc. Or am I missing something?

Reply:
Actually,  I am using TLS.
```c++
            if (!p_element) {

                std::stringstream name;
                const int thread_id = OpenMPUtils::ThisThread();
                name << rModelPart.Name() << "_temp_" << thread_id;
                GeometryType::PointsArrayType nodes;
                {
                    KRATOS_CRITICAL_SECTION

                    model_part_names.push_back(name.str());
                    auto& tls_model_part = rModelPart.CreateSubModelPart(name.str());

                    for (IndexType i = 0; i < r_geometry.size(); ++i) {
                        nodes.push_back(tls_model_part.CreateNewNode((thread_id + 1) * max_id + i + 1, r_geometry[i]));
                    }
                }

                p_element = rElement.Create(1, nodes, rElement.pGetProperties());
                p_element->Initialize(r_process_info);
                p_element->InitializeSolutionStep(r_process_info);
            }
```
The above code segment is only executed **once per thread, and it only adds 4 nodes for the model part** hence, it is not expensive. I create the model part because, in kratos it is not advised to create `Nodes` on the fly because, the `Node` does not hold the `SolutionStepVariablesList`, it owned by the model part. So if you create `Nodes` without model part, then you may encounter seg-faults which are hard to trace.